compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -2802,14 +2802,32 @@ abstract class GenJSCode extends plugins.PluginComponent
          *   {name1: arg1, name2: arg2, ... }
          */
 
+        def warnIfDuplicatedKey(keys: List[String], namesSeen: Set[String] = Set.empty): Unit = keys match {
```

Comment:
Add `@tailrec`.

Also, stay under the 80 character limit if possible. Here you can easily break before `namesSeen` (and indent 4 spaces, since it is inside parentheses). (This means you'll have to put `keys match {` on the next line, inside a block, otherwise it will look weird.)